File: 8_heap/clampable/ClampableHeap.ts
import { Heap } from '../Heap'

type Pair = { value: number; count: number }

class ClampableHeap {
  private readonly _clampMin: boolean
  private readonly _heap: Heap<Pair>
  private _total = 0
  private _count = 0

  /**
   * @param clampMin 为true时，支持容器内所有数与x取min；为false时，支持容器内所有数与x取max.
   */
  constructor(clampMin: boolean) {
    this._clampMin = clampMin
    this._heap = this._clampMin
      ? new Heap<Pair>((a, b) => b.value - a.value)
      : new Heap<Pair>((a, b) => a.value - b.value)
  }

  add(x: number): void {
    this._heap.push({ value: x, count: 1 })
    this._total += x
    this._count++
  }

  clamp(x: number): void {
    let newCount = 0
    if (this._clampMin) {
      while (this._heap.size > 0) {
        const top = this._heap.peek()!
        if (top.value < x) break
        this._heap.pop()
        this._total -= top.value * top.count
        newCount += top.count
      }
    } else {
      while (this._heap.size > 0) {
        const top = this._heap.peek()!
        if (top.value > x) break
        this._heap.pop()
        this._total -= top.value * top.count
        newCount += top.count
      }
    }
    this._total += x * newCount
    this._heap.push({ value: x, count: newCount })
  }

  sum(): number {
    return this._total
  }

  get size(): number {
    return this._count
  }
}

export { ClampableHeap }

if (require.main === module) {
  const pq1 = new ClampableHeap(true)
  pq1.add(1)
  pq1.add(2)
  pq1.add(3)
  console.assert(pq1.sum() === 6)
  pq1.clamp(2)
  console.assert(pq1.sum() === 5)
  pq1.clamp(1)
  pq1.add(2)
  console.assert(pq1.sum() === 5)

  const pq2 = new ClampableHeap(false)
  pq2.add(1)
  pq2.add(2)
  pq2.add(3)
  console.assert(pq2.sum() === 6)
  pq2.clamp(2)
  console.assert(pq2.sum() === 7)
  pq2.clamp(3)
  pq2.add(2)
  console.assert(pq2.sum() === 11)
}

File: 8_heap/clampable/ClampableHeap.py
from heapq import heappop, heappush


class ClampableHeap:
    __slots__ = ("_clampMin", "_total", "_count", "_heap")

    def __init__(self, clampMin: bool):
        """
        clampMin:
          为true时,调用Clamp(x)后,容器内所有数最小值被截断(小于x的数变成x);
          为false时,调用Clamp(x)后,容器内所有数最大值被截断(大于x的数变成x).
          如果需要同时支持两种操作,可以使用双端堆.
        """
        self._clampMin = clampMin
        self._total = 0
        self._count = 0
        self._heap = []

    def add(self, x: int) -> None:
        heappush(self._heap, (x if self._clampMin else -x, 1))
        self._total += x
        self._count += 1

    def clamp(self, x: int) -> None:
        newCount = 0
        if self._clampMin:
            while self._heap:
                v, c = self._heap[0]
                if v > x:
                    break
                heappop(self._heap)
                self._total -= v * c
                newCount += c
            self._total += x * newCount
            heappush(self._heap, (x, newCount))
        else:
            while self._heap:
                v, c = self._heap[0]
                v = -v
                if v < x:
                    break
                heappop(self._heap)
                self._total -= v * c
                newCount += c
            self._total += x * newCount
            heappush(self._heap, (-x, newCount))

    def sum(self) -> int:
        return self._total

    def __len__(self) -> int:
        return self._count


if __name__ == "__main__":
    ch = ClampableHeap(clampMin=False)
    ch.add(1)
    ch.add(2)
    ch.add(3)
    assert ch.sum() == 6
    ch.clamp(2)
    assert ch.sum() == 5
    ch.clamp(1)
    ch.add(2)
    assert ch.sum() == 5

    ch = ClampableHeap(clampMin=True)
    ch.add(1)
    ch.add(2)
    ch.add(3)
    assert ch.sum() == 6
    ch.clamp(2)
    assert ch.sum() == 7
    ch.clamp(3)
    ch.add(2)
    assert ch.sum() == 11
